(wip) tutorial/parser

anpaul0615 opened the pull request.

anpaul0615 committed:
chore: 코드 구조 작성

anpaul0615 commented:
### 기존 코드 검토 및 실행흐름 정리 ([https://github.com/githru/git-metadata-preprocessor](https://github.com/githru/git-metadata-preprocessor))
- main log 파싱        
    ```jsx
    "commit" 포함
    => 아이디(auto_inc), 커밋해시, 부모커밋해시, diffStat, message
    
    "Merge" 포함
    => 브랜치해시1, 브랜치해시2
    
    "Author:" 포함
    => author
    
    "AuthorDate:" 포함
    => authorDate
    
    "Committer:" 포함
    => comitter
    
    "CommitDate:" 포함
    => comitDate
    
    else
      "Signed-off-by:" 포함
      =>
    
      "\t" split
      => 추가된라인, 삭제된라인, 파일명
      => diffStat 누적입력
    
      나머지는 message 에 포함
    ```
    
- ref log 파싱
    
    ```jsx
    "tag:" 포함
    ⇒ tag 할당
    
    "HEAD -> " 포함
    ⇒ isHead: true 할당
    
    else
    ⇒ branches 할당
    ⇒ tags 할당
    ```


### git log parse 참고자료 리서치
- [https://git-scm.com/docs/pretty-formats](https://git-scm.com/docs/pretty-formats)
    
    ![image](https://user-images.githubusercontent.com/8488507/181920193-11419ca9-aa6a-4a0d-8e29-8f18f1bd6617.png)
    
    => git-metadata-preprocessor 에서 사용하는 git log 포맷이 `--pretty=fuller` 
    
    => 상기 포맷대로 정규식 작성해서 필요정보 추출
    
    ```jsx
    /* git log --help */
    
    //...
    
    Diff Formatting
       Listed below are options that control the formatting of diff output. Some of them are specific to
       git-rev-list(1), however other diff options may be given. See git-diff-files(1) for more options.
    
       -c
           With this option, diff output for a merge commit shows the differences from each of the
           parents to the merge result simultaneously instead of showing pairwise diff between a parent
           and the result one at a time. Furthermore, it lists only files which were modified from all
           parents.
    
       --cc
           This flag implies the -c option and further compresses the patch output by omitting
           uninteresting hunks whose contents in the parents have only two variants and the merge result
           picks one of them without modification.
    
       --combined-all-paths
           This flag causes combined diffs (used for merge commits) to list the name of the file from all
           parents. It thus only has effect when -c or --cc are specified, and is likely only useful if
           filename changes are detected (i.e. when either rename or copy detection have been requested).
    
       -m
           This flag makes the merge commits show the full diff like regular commits; for each merge
           parent, a separate log entry and diff is generated. An exception is that only diff against the
           first parent is shown when --first-parent option is given; in that case, the output represents
           the changes the merge brought into the then-current branch.
    
       -r
           Show recursive diffs.
    
       -t
           Show the tree objects in the diff output. This implies -r.
    ```
    
    ⇒ git log `-c` 옵션
    
- [https://github.com/wayfair/git-parse](https://github.com/wayfair/git-parse)
    
    ⇒ 최근까지 메인테이닝 되고있는 git parse 관련 라이브러리면서, 코드 곳곳에 채용할만한 좋은 아이디어가 많이보여서 선정
    
    ```jsx
    git log --pretty=format:@@BEGIN@@%n%H%n%an%n%ae%n%aD%n@@MESSAGE@@%n%B%n@@END@@ --name-status
    ```
    
    ⇒ 파싱할 대상을 `@@BEGIN@@` `@@MESSAGE@@` `@@END@@` 형식으로 git log 텍스트 포멧에 마킹한다음, 마커를 기준으로 텍스트 추출파싱하는 방식 사용
    
    ```jsx
    const gitLogStream = (...) => {
     const gitProcess = childProcess.spawn("git", gitParams, { cwd: pathToRepo });
     // ...
     return {
       stream: gitProcess.stdout,
       addErrorHandler: (fn: GitLogStreamErrorHandler) => errorHandlers.push(fn),
     };
    }
    ```
    
    ```jsx
    const gitToJs = (...) => {
      const { stream, addErrorHandler } = gitLogStream(resolvedPath, options);
      //...
      return new Promise((resolve, reject) => {
        // ...
        parseGitLog(stream).then(resolve);
      });
    };
    ```
    
    ⇒ git log 뽑아내는 함수에서 stdout 스트림 그대로 반환 + 스트림 그대로 파싱함수에 전달 
    
    ```jsx
    const parseGitLog = (stream: NodeJS.ReadableStream): Promise<GitCommit[]> => {
      return new Promise((resolve, reject) => {
        // ...
        const streamByLine = byline(stream);
    
        streamByLine.on("data", (line) => {
          const lineString = line.toString();
          // ...
          buffer.push(lineString);
        });
    
        streamByLine.on("error", (e) => {
          reject(e);
        });
    
        streamByLine.on("end", () => {
          if (buffer.length) {
            parsedCommits.push(parseCommit(buffer));
          }
          resolve(parsedCommits);
        });
      });
    };
    ```
    
    ```jsx
    const parseCommit = (commit: string[]): GitCommit => {
      // ...
      const parsedCommit = {
        hash,
        authorName,
        authorEmail,
        date,
        message,
        filesAdded,
        filesDeleted,
        filesModified,
        filesRenamed,
      };
      return parsedCommit;
    };
    ```
    
    ⇒ 받아온 스트림을 line by line 으로 toString 해서 모아놓은다음 + 한방에 파싱함수 호출
    
    ⇒ check: 대량의 text stream 을 한번에 모으면 nodejs 기본메모리 2gb 를 초과할수도 있음
    
    ⇒ todo: 파싱함수를 line by line 으로 실행하고, 그결과를 writestream 으로 써내려가면 될것같음
    
- [https://github.com/domharrington/node-gitlog](https://github.com/domharrington/node-gitlog)
    
    ```jsx
    const fieldMap = {
      hash: "%H",
      abbrevHash: "%h",
      treeHash: "%T",
      abbrevTreeHash: "%t",
      parentHashes: "%P",
      abbrevParentHashes: "%P",
      authorName: "%an",
      authorEmail: "%ae",
      authorDate: "%ai",
      authorDateRel: "%ar",
      committerName: "%cn",
      committerEmail: "%ce",
      committerDate: "%cd",
      committerDateRel: "%cr",
      subject: "%s",
      body: "%b",
      rawBody: "%B",
      tag: "%D",
    } as const;
    ```
    
    ⇒ git log 옵션과 필드를 각각 매핑해서 코드레벨에서 제어할수있도록 구성

anpaul0615 commented:
### 구현 아이디어
- `git log --all --no-pager --parents --numstat --date-order --pretty=fuller` 를 정규식으로 파싱
- `git log ~~~` 를 stream 으로 받아옴 + line 별로 파싱함수 호출 + commits.json 생성


### 고민거리: commits.json 에 append 하는 방식
- json 을 만드는 처리가 중간에 실패할때 (ex. 사용자가 vscode 강제종료)
  1) commits.json 생성이 아예 실패하거나
  2) commits.json 이 불완전한 json 으로 write 되거나
- commits.json 이 아예 실패하게되면, 처음부터 다시시작..?
- 그것보다는… commits.json 에 write append 하는게 좋을것같지만.. 사실이것도 json valid 하게 만들기위한 추가비용이 필요함
- 그러면 commit 단위로 commit.[hash].chunk.json 형식으로 저장하고, 파싱이 완료된 이후에 commit.chunk.json 을 commits.json 으로 바꿔주는 방식 사용..?
- 이거보다 좀더 좋은방법은.. 차라리 sqlite 를 사용하는게 좋을지도..?

ytaek commented:
> ### 고민거리: commits.json 에 append 하는 방식
> * json 을 만드는 처리가 중간에 실패할때 (ex. 사용자가 vscode 강제종료)
>   
>   1. commits.json 생성이 아예 실패하거나
>   2. commits.json 이 불완전한 json 으로 write 되거나
> * commits.json 이 아예 실패하게되면, 처음부터 다시시작..?
> * 그것보다는… commits.json 에 write append 하는게 좋을것같지만.. 사실이것도 json valid 하게 만들기위한 추가비용이 필요함
> * 그러면 commit 단위로 commit.[hash].chunk.json 형식으로 저장하고, 파싱이 완료된 이후에 commit.chunk.json 을 commits.json 으로 바꿔주는 방식 사용..?
> * 이거보다 좀더 좋은방법은.. 차라리 sqlite 를 사용하는게 좋을지도..?

여기에 대해서 저도 좀 고민이 많은데...
일단 시작은 그냥 full log를 훑고 가는 것처럼 하려고 합니다.
(Githru때 commit이 2만개 정도 되는 realm-java를 샘플로 할 때, json 파일 크기가 그리 엄청나게 크지는 않았거든요)

stream 형식으로 받아오는 부분은 좀 뒤에 고민해 보는걸루...

git이 history를 permanently 보존하거나, 못 바꾸게 하는게 아니라서, 요 부분은 좀 다양한 각도로 고민이 필요할 것 같아요.


정말 심도 깊은 내용들 감사합니다!!

anpaul0615 commented:
> 일단 시작은 그냥 full log를 훑고 가는 것처럼 하려고 합니다. (Githru때 commit이 2만개 정도 되는 realm-java를 샘플로 할 때, json 파일 크기가 그리 엄청나게 크지는 않았거든요)
> 
> stream 형식으로 받아오는 부분은 좀 뒤에 고민해 보는걸루...

미처 정제하지못한 고민거리에 친절한 답변(?) 해주셔서 감사합니다ㅎㅎㅎ
말씀대로 우선 full log 뽑아서 파싱진행하는 방식으로 만들어볼게요!

anpaul0615 committed:
chore: jest 환경설정

anpaul0615 committed:
feat: git log 수집기 구현

anpaul0615 commented:
`git-repos/` 디렉토리 아래에 위치한 레포지토리의 git log 출력을 text 로 받아오는 기능을 구현해봤습니다.

구현된 기능은 아래와 같이 동작합니다.

```
// cd <WORKSPACE>/parser/anpaul0615/git-repos
// git clone <GIT-REPO-URI>

const repoName = '<GIT-REPO>';
const gitLog = await collect(repoName);
console.log(gitLog)

// do something...
```

ytaek commented:
PR에 덧붙이면 approval 받은게 날라가기도 하고, 재review요청 알람도 안 가는 것 같아요.
어느정도 완료하고 승인받은 코드의 후속작업은 PR은 새롭게 다시 만들어서 해도 괜찮겠습니다~.

anpaul0615 committed:
feat: git log 분석기 구현

anpaul0615 committed:
refactor: git log 분석기 개선

- 내부 파싱처리를 좀 더 세밀하게 구분

- ASIS
  - git log 출력을 PrettyFuller 포맷 목록으로 파싱
  - PrettyFuller 포맷을 Githru 포맷으로 파싱

- TOBE
  - parseGitLogToCommitUnit : git log 출력을 커밋단위 목록으로 파싱
  - parseCommitUnitToPrettyFullerCommit : 각 커밋단위의 header/message/numStat text를 PrettyFuller 포맷으로 파싱
  - parsePrettyFullerCommitToGithruCommit : PrettyFuller 포맷 text를 Githru 포맷으로 파싱

anpaul0615 committed:
feat: commits.json 저장기 구현

anpaul0615 committed:
fix: commits.json 저장기 동작방식 변경

- ASIS
  - 이미 생성된 commits.json 있으면 에러

- TOBE
  - 이미 생성된 commits.json 있으면 덮어쓰기

anpaul0615 commented:
> PR에 덧붙이면 approval 받은게 날라가기도 하고, 재review요청 알람도 안 가는 것 같아요.
> 어느정도 완료하고 승인받은 코드의 후속작업은 PR은 새롭게 다시 만들어서 해도 괜찮겠습니다~.

코멘트를 늦게 확인해서 커밋이 올라가버렸습니다..ㅠㅠ

이왕 이렇게된거 잠깐 자기변호(?)를 좀 하자면... 
천천히 생각하면서 구현 진행하려고 처음부터 한 PR에 누적해서 커밋 올리려고 했습니다.
한단계 한단계씩 커밋 올릴때마다 여기 PR에서 의견을 나누는(?) 방식을 생각했어요.

어차피 이 PR 의 목적은 parser 를 완전하게 구현하는게 목적이니,
그 과정중에 생긴 approve 는 크게 의미가 없다고 생각했고,
최종적으로는 기능구현 끝난다음에 approve 받는게 목표였어요.

그런데 가만 생각해보니깐, 그럴거면 차라리 특정 피처에 대한 구현 과정 자체를 이슈에서 논의하는게 더 좋을것같았어요. (애초에 그러라로 만든 기능인것같기도하구요) 이 부분은 제가 github 을 잘못 사용한 것 같아요... 😭

다음부터는 개발방향 논의는 이슈 생성해서 진행하도록 할게요..! 좋은 의견 감사합니다.

ansrlm commented on parser/anpaul0615/src/repository.ts:
```diff
@@ -0,0 +1,36 @@
+import * as fs from "fs";
```
fs 모듈이 메소드가 생각보다 적지 않은 것으로 기억하는데 필요한 부분만 가져와도 좋을 것 같습니다~

anpaul0615 replied:
넵넵 다음 커밋에 참고하겠습니다. 감사합니다 :)

ansrlm reviewed (approved):
고생하셨습니다.
LGTM 👍

ytaek commented on parser/anpaul0615/src/parser.ts:
```diff
@@ -0,0 +1,85 @@
+const CommitterPattern =  /Commit:     (?<commiter>.+) <(?<commiterEmail>.+)>/;
+const CommitDatePattern = /CommitDate: (?<commitDate>.+)/;
+
+const PrettyFullerHeaderPattern = new RegExp([
+  CommitPattern,
+  /\n/,
+  MergePattern,
+  AuthorPattern,
+  /\n/,
+  AuthorDatePattern,
+  /\n/,
+  CommitterPattern,
+  /\n/,
+  CommitDatePattern,
+].map((r) => r.source).join(""), "m");
+
```
👍

anpaul0615 replied:
: )

ytaek reviewed (approved):
LGTM!
